### PR TITLE
Adding CUB sort to our GPU sorting abilities

### PR DESCRIPTION
**Pull request by ShreyasKhandekar**
This PR adds the ability to call out to CUB's Device radix sort: https://nvidia.github.io/cccl/cub/api/structcub_1_1DeviceRadixSort.html? from within our GPU sorting functionality.

The `gpuSort` function now prefers this much faster sort whenever it is supported. This improves our sorting on GPU's drastically on both Nvidia and AMD GPUs (with some restrictions for AMD based on the ROCm version)

While there this PR also does the following:
* The `gpuSort` function should also work in CPU-as-device mode now.
* The in house `parallelRadixSort` now supports all `uint` coercible types
* Clean up the runtime CUB interface and rename some things for better code maintainability
* Add a gpu sort performance explorer program to be used for nice sorting explorations on different problem sets and sizes
* Fix a memory leak in CUB based reductions

Testing:
- [x] nvidia AOD
- [x] nvidia UM
- [x] AMD AOD
- [x] AMD UM
- [x] CPU-as-device

### CODE REVIEW

File: modules/standard/GPU.chpl [hunk 1]
```diff
@@ -910,16 +952,33 @@ module GPU
...
+
+  private proc fallBackRadixSort(ref gpuInputArr : [] ?t) where t==uint{
     // Based on the inputArr size, get a chunkSize such that numChunks is on the order of thousands
     // TODO better heuristic here?
     var chunkSize = Math.divCeil(gpuInputArr.size, 2000);
     parallelRadixSort(gpuInputArr, bitsAtATime=8, chunkSize, noisy=false, distributed=false);
   }
 
+  private proc fallBackRadixSort(ref gpuInputArr : [] ?t) where t!=uint{
+    compilerError("Womp Womp");
```
Comment (mppf) [inline modules/standard/GPU.chpl:980]: lol but you have to make a better error here

File: modules/standard/GPU.chpl [hunk 2]
```diff
@@ -895,6 +895,55 @@ module GPU
 
   private import Time;
 
+  // We no doc it so we can test this independently
```
Comment (e-kayrakli) [inline modules/standard/GPU.chpl:898]: "... as opposed to making it `private`" I presume?

File: modules/standard/GPU.chpl [hunk 3]
```diff
@@ -895,6 +895,55 @@ module GPU
...
+    }
+
+    // Only useful when calling gpuCubSort directly
+    extern proc chpl_gpu_can_cub_sort(): bool;
+    if !chpl_gpu_can_cub_sort() {
+      gpuSort(gpuInputArr);
+      return;
+    }
+
+    proc getExternFuncName(param op: string, type t) param: string {
```
Comment (e-kayrakli) [inline modules/standard/GPU.chpl:925]: Looks like you can drop `t`

File: modules/standard/GPU.chpl [hunk 4]
```diff
@@ -895,6 +895,55 @@ module GPU
...
+                      else if t==real(64) then "double"
+                      else                     "unknown";
+
+    if cTypeName == "unknown" {
+      compilerError("Arrays with ", t:string,
+                    " elements cannot be sorted with gpuCubSort functions");
+    }
+
+    // Only useful when calling gpuCubSort directly
+    extern proc chpl_gpu_can_cub_sort(): bool;
```
Comment (e-kayrakli) [inline modules/standard/GPU.chpl:919]: Let's not put `cub` in the name. I see that you probably did that because GPU can sort with or without CUB because of the in-house implementation we have. But I don't think that holds much water for me. The api name could change down the road for NVIDIA and AMD. When we have support for Intel, whose software stack is not CUDA-inspired I am sure `cub` will not make much sense.

File: modules/standard/GPU.chpl [hunk 5]
```diff
@@ -895,6 +895,55 @@ module GPU
...
+    var temp : gpuInputArr.type;
+
+    const basePtr = c_ptrToConst(gpuInputArr);
+    const tempPtr = c_ptrTo(temp);
+    // Do the sort
+    sort_fn(basePtr, tempPtr, gpuInputArr.size);
+
+    // The sorted values are in temp, so we need to copy them back
+    // to the original array
+    gpuInputArr = temp;
```
Comment (e-kayrakli) [inline modules/standard/GPU.chpl:944]: Hmm... can we do a `<=>` here? As much as it seems to be a redundant operation, what you have here will be a memcpy of the full data, whereas since `gpuInputArr` and `temp` are of same type and DefaultRectangular, I expect `<=>` to be just a pointer swap of implementation classes or the data buffer or something.

But premature optimization is root of all evil, so, maybe do that only if it helps with performance? And if it doesn't (or if you're short on time), maybe just add a comment here hinting at a potential optimization down the road?

File: modules/standard/GPU.chpl [hunk 6]
```diff
@@ -910,21 +959,46 @@ module GPU
...
+
+  private proc fallBackRadixSort(ref gpuInputArr : [] ?t) where isCoercible(t, uint){
     // Based on the inputArr size, get a chunkSize such that numChunks is on the order of thousands
     // TODO better heuristic here?
     var chunkSize = Math.divCeil(gpuInputArr.size, 2000);
     parallelRadixSort(gpuInputArr, bitsAtATime=8, chunkSize, noisy=false, distributed=false);
   }
 
+  private proc fallBackRadixSort(ref gpuInputArr : [] ?t) where !isCoercible(t, uint){
+    compilerError("GPU Based sorting without CUB is only supported for arrays of type uint. Please check your ROCM Version and make sure it's >= 5.0.0");
```
Comment (e-kayrakli) [inline modules/standard/GPU.chpl:989]: This is independent of CUB, is it? Is there any chance we can fall here except with ROCm 4?

File: modules/standard/GPU.chpl [hunk 7]
```diff
@@ -910,21 +959,46 @@ module GPU
...
+  }
   // We no doc it so we can test this independently to simulate all cases that can happen with sort
   @chpldoc.nodoc
-  proc parallelRadixSort(ref gpuInputArr : [] uint, const bitsAtATime : int = 8, const chunkSize : int = 512, const noisy : bool = false,
-                         const distributed : bool = false) { // The last argument is for multi GPU sort that is pending a patch before it can work
-
+  proc parallelRadixSort(ref gpuInputArr : [] ?t, const bitsAtATime : int = 8, const chunkSize : int = 512, const noisy : bool = false,
+                         const distributed : bool = false) where isCoercible(t, uint){ // The last argument is for multi GPU sort that is pending a patch before it can work
```
Comment (e-kayrakli) [inline modules/standard/GPU.chpl:994]: could you reflow this and `fallBackRadixSort` to fit in 80 cols (prefer) or at least 100 (can live with it)?

File: runtime/include/chpl-gpu-impl.h [hunk 8]
```diff
@@ -74,6 +74,7 @@ bool chpl_gpu_impl_stream_ready(void* stream);
 void chpl_gpu_impl_stream_synchronize(void* stream);
 
 bool chpl_gpu_impl_can_reduce(void);
+bool chpl_gpu_impl_can_cub_sort(void);
```
Comment (e-kayrakli) [inline runtime/include/chpl-gpu-impl.h:77]: again, let's not have "cub" in the interface. (similar to what we already have for reduce)

File: runtime/include/chpl-gpu-impl.h [hunk 9]
```diff
@@ -87,6 +88,15 @@ GPU_REDUCE(DECL_ONE_REDUCE_IMPL, maxloc)
 
 #undef DECL_ONE_REDUCE_IMPL
 
+#define DECL_ONE_SORT_IMPL(chpl_kind, data_type) \
+void chpl_gpu_impl_sort_##chpl_kind##_##data_type(data_type* data_in, \
+                                                  data_type* data_out, \
+                                                  int n, void* stream);
+GPU_SORT_TYPES(DECL_ONE_SORT_IMPL, keys)
+// TODO: GPU_SORT(DECL_ONE_SORT_IMPL, keysDesc/ DoubleBuffer/Pairs etc)
```
Comment (e-kayrakli) [inline runtime/include/chpl-gpu-impl.h:96]: cruft? Otherwise, I am not following what this TODO is about.

File: runtime/include/chpl-gpu.h [hunk 10]
```diff
@@ -171,6 +171,7 @@ bool chpl_gpu_can_access_peer(int dev1, int dev2);
 void chpl_gpu_set_peer_access(int dev1, int dev2, bool enable);
 
 bool chpl_gpu_can_reduce(void);
+bool chpl_gpu_can_cub_sort(void);
```
Comment (e-kayrakli) [inline runtime/include/chpl-gpu.h:174]: no "cub"

File: runtime/include/gpu/chpl-gpu-reduce-util.h [hunk 11]
```diff
@@ -43,5 +43,29 @@
...
+  MACRO(chpl_kind, int8_t)  \
+  MACRO(chpl_kind, int16_t)  \
+  MACRO(chpl_kind, int32_t)  \
+  MACRO(chpl_kind, int64_t)  \
+  MACRO(chpl_kind, uint8_t)  \
+  MACRO(chpl_kind, uint16_t)  \
+  MACRO(chpl_kind, uint32_t)  \
+  MACRO(chpl_kind, uint64_t)  \
+  MACRO(chpl_kind, float)   \
+  MACRO(chpl_kind, double);
```
Comment (e-kayrakli) [inline runtime/include/gpu/chpl-gpu-reduce-util.h:68]: It looks to me like we could have reused the ones above. Can we do that?

You can use `CUB` here in the macro names as this is way too much detail -- `GPU_CUB_WRAP`, `GPU_DEV_CUB_WRAP` or something?

File: runtime/src/gpu/amd/gpu-amd.c [hunk 12]
```diff
@@ -378,4 +378,8 @@ bool chpl_gpu_impl_can_reduce(void) {
   return ROCM_VERSION_MAJOR>=5;
 }
 
+bool chpl_gpu_impl_can_cub_sort(void){
+  return chpl_gpu_impl_can_reduce();
```
Comment (e-kayrakli) [inline runtime/src/gpu/amd/gpu-amd.c:382]: `can_sort`

File: runtime/src/gpu/nvidia/gpu-nvidia-reduce.cc [hunk 13]
```diff
@@ -81,5 +81,30 @@ GPU_IMPL_REDUCE(DEF_ONE_REDUCE_RET_VAL_IDX, ArgMax, maxloc)
 
 #undef DEF_ONE_REDUCE_RET_VAL_IDX
 
+#define DEF_ONE_SORT(cub_kind, chpl_kind, data_type) \
```
Comment (e-kayrakli) [inline runtime/src/gpu/nvidia/gpu-nvidia-reduce.cc:84]: we should probably rename this file and its AMD counterpart. Again "cub" is OK to use in the name, these files are literally cub/hipcub wrappers.

File: test/gpu/native/studies/sort/gpu-sort-performance-explorer.chpl [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+      on here.gpus[0]{
+        var arr = input; // Copy to gpu
+        t.start();
+        testsort(arr, m);
+        t.stop();
+        input = arr; // Copy back to cpu
+      }
+    }
+    var mibs = mibibytes * ntrials / t.elapsed();
+    // if printStats then writef(" % 12s sec  ", t.elapsed():string);
```
Comment (e-kayrakli) [inline test/gpu/native/studies/sort/gpu-sort-performance-explorer.chpl:155]: Maybe uncomment?

File: test/gpu/native/studies/sort/gpu-sort-performance-explorer.chpl [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+  }
+  if printStats then
+    writeln();
+
+  var size = minSize;
+  while size <= maxSize {
+    testsize(size);
+    size *= 2;
+  }
+}
```
Comment (e-kayrakli) [inline test/gpu/native/studies/sort/gpu-sort-performance-explorer.chpl:184]: I haven't reviewed this file in detail. If you have parts that you need me to look at, I'd be happy to.

**Review by e-kayrakli — APPROVED** [review]
Great work!

I have some inline comments for some changes I wish to see before this could be merged, but they are more about code maintainability then anything else.

File: modules/standard/GPU.chpl (hunk 4 above)
Comment (ShreyasKhandekar) [reply modules/standard/GPU.chpl:919]: Yeah I put CUB in the name because the function tells us whether we can use CUB for the sort or not. 
I can rename this to `chpl_gpu_can_extern_sort` and rename `gpuCubSort` to gpuExternSort` as well if you'd like.

File: modules/standard/GPU.chpl (hunk 5 above)
Comment (ShreyasKhandekar) [reply modules/standard/GPU.chpl:944]: I'll add a comment for now just in case it causes issues in testing which I won't have time to resolve

File: modules/standard/GPU.chpl (hunk 6 above)
Comment (ShreyasKhandekar) [reply modules/standard/GPU.chpl:989]: From what I understand, the `fallBackRadixSort` is only there for ROCm 4. All CUDA versions will use the CUB sort and CPU-as-device mode will use the CPU sort. I mention CUB because the limitation of only using `uint`s arises from the absence of CUB (and because that's what the in house implementation supports).

Does this sound better:
```suggestion
    compilerError("GPU Based sorting is only supported for arrays of type uint for ROCm version <5.0.0. Upgrade your ROCm install to sort all primitive numeric types");
```

File: runtime/include/chpl-gpu-impl.h (hunk 9 above)
Comment (ShreyasKhandekar) [reply runtime/include/chpl-gpu-impl.h:96]: Had to google cruft haha.
This TODO is just to demonstrate that there are more functions under cub::DeviceRadixSort that can be added later. These include decending sorts, DoubleBufferSorts (which use less auxillary space), and SortPairs which can sort key value pairs, and also potentially be extended to sort any type.

File: runtime/include/gpu/chpl-gpu-reduce-util.h (hunk 11 above)
Comment (ShreyasKhandekar) [reply runtime/include/gpu/chpl-gpu-reduce-util.h:68]: I could reuse the ones above, but it would make it a little harder for someone to follow the code by reading it (it's already pretty hard)

I'm not sure what you mean by too much detail and what you are suggesting I rename. Also I should use CUB in the names here but not in the other code? I guess I'm not following your naming philosophy here.

File: runtime/src/gpu/amd/gpu-amd.c (hunk 12 above)
Comment (ShreyasKhandekar) [reply runtime/src/gpu/amd/gpu-amd.c:382]: I think `chpl_gpu_impl_can_sort` and `chpl_gpu_can_sort` make it sound like it checks if our in house sort can sort it as well. 
I would want to make a distinction between the two. I suggested `chpl_gpu_impl_can_extern_sort` above, `chpl_gpu_impl_can_runtime_sort` is also an option

File: test/gpu/native/studies/sort/gpu-sort-performance-explorer.chpl (hunk 15 above)
Comment (ShreyasKhandekar) [reply test/gpu/native/studies/sort/gpu-sort-performance-explorer.chpl:184]: I forgot to add a comment to the top of this file saying it's adapted from Michael's `test/library/packages/Sort/performance/sort-performance-explorer.chpl` 

I don't think it needs an in depth review :)

File: runtime/include/gpu/chpl-gpu-reduce-util.h (hunk 11 above)
Comment (ShreyasKhandekar) [reply runtime/include/gpu/chpl-gpu-reduce-util.h:68]: Okay I stared at it and I see what you meant (hopefully):
- Rename file to `chpl-gpu-cub-util.h`
- Rename `GPU_IMPL_REDUCE` to `GPU_DEV_CUB_WRAP` and `GPU_REDUCE` to `GPU_CUB_WRAP`
- Remove `GPU_IMPL_SORT_TYPES` and `GPU_SORT_TYPES`

File: modules/standard/GPU.chpl (hunk 4 above)
Comment (e-kayrakli) [reply modules/standard/GPU.chpl]: I recommend just `chpl_gpu_can_sort`. It is a runtime function and I view it as asking the runtime whether it can sort it.

I can tolerate `chpl_gpu_can_extern_sort`, if you wanna argue for it more strongly. My reasons for not liking it:

- Doesn't add much in terms of meaning in my view
- Is Longer
- Is different than `chpl_gpu_can_reduce` (obviously, if you want to stick with `chpl_gpu_can_extern_sort`, you should rename the reduce support functions, too)

File: runtime/include/chpl-gpu-impl.h (hunk 9 above)
Comment (e-kayrakli) [reply runtime/include/chpl-gpu-impl.h:96]: Can you clarify that in the comment? Alternatively, you can drop the comment and create a private issue, too.

File: runtime/src/gpu/amd/gpu-amd.c (hunk 12 above)
Comment (e-kayrakli) [reply runtime/src/gpu/amd/gpu-amd.c:382]: We can have the name discussion in the thread above.

Sorry about posting an unhelpful comment. This line calls `chpl_gpu_impl_can_reduce`. It should call the `_sort` version.

File: modules/standard/GPU.chpl (hunk 6 above)
Comment (e-kayrakli) [reply modules/standard/GPU.chpl]: I'm good with that.

File: runtime/include/gpu/chpl-gpu-reduce-util.h (hunk 11 above)
Comment (e-kayrakli) [reply runtime/include/gpu/chpl-gpu-reduce-util.h]: Yeah, I like those changes. (Again, sorry for quick/terse comments).

**Comment by e-kayrakli** [comment]
Since I got you confused by my naming opinions, I'll just describe them here. I don't think we are in a big disagreement, but I think I owe you more words. :)

Modularity and good abstractions make more maintainable software.

I see our runtime as an entity of its own. The module code can ask the runtime whether it can sort or not, and based on that can fall back to an implementation independent of the runtime (completely in Chapel-land, or directly calling into a non-runtime extern). The runtime interface (`chpl-gpu.h`) shouldn't include `cub` in the name, because the module code doesn't need to know or care whether it is based on CUB or some other library. Runtime should completely abstract away what it uses under the hood (unless there are alternatives where we want the module code to make a decision). This is the philosophical side of the story.

More practically, adding `cub` in the name in the runtime interface is extra maintainability burden -- what if we change the library we use for sorting in the future? What if we implement our own device runtime that's capable of sorting (imagine a `__device__` function in the runtime that can sort faster than CUB for some specific data type)? My last rhetorical question kinda implies but, one can even imagine runtime relying on multiple strategies (distributed arrays? Some other library for custom types?) for sorting down the road. We will likely have to do something non-CUB we target Intel for example. I do see that the burden of renaming it then would be probably minimal, but having "cub" in the runtime interface just feels wrong.

You proposed other alternatives like `can_extern_sort` or `can_runtime_sort`. I am game with those. I wouldn't (and didn't with `chpl_gpu_can_reduce` which admittedly doesn't have a module-code fallback) because it is too wordy. But it is not something I would fight about. I would just ask that you make the same update to `chpl_gpu_can_reduce` with either of those wordings for consistency. (Prrrrobably `can_extern_sort` is better as `runtime` in a runtime API feels tautological)

As for being OK with other "CUB"s in the implementation: macros and files where I think CUB appear in the name are the points where we interact with CUB directly. At that point, I don't think we need to abstract away the fact that we are working with CUB, and if it helps, we can use it in the names.

I'd be happy to clarify further.

File: runtime/src/gpu/amd/gpu-amd.c (hunk 12 above)
Comment (ShreyasKhandekar) [reply runtime/src/gpu/amd/gpu-amd.c:382]: I've made it call `chpl_gpu_impl_can_reduce` on purpose because both of them have the same check needed: `ROCM_VERSION_MAJOR>=5`. So instead of duplicating the check, I just call the function above it

**Comment by ShreyasKhandekar** [comment]
Thanks for the clarification Engin! I agree with you completely. I guess having `extern` in the name of an already `extern` function doesn't really add much. I will just go with `chpl_gpu_can_sort`
